rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,117 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a magic comment in the first line of a Nix file:
+
+    #? Nix <version>
+
+where `<version>` is a released version of Nix the given file is intended to work with.
+
+If no magic comment is present, Nix 2.12 – the stable release at the time this RFC was accepted – is assumed.
+
+Add an appropriate parameter to commands that allow encoding the same information where where files are not involved.
+
```

Comment:
This RFC should also discuss the interaction with flakes. Probably we want the version tag on the top-level `flake.nix` to be inherited by all the `.nix` files in the flake, so you don't have to specify the version for every file in the flake.

Reply:
Previously also briefly discussed [here](https://github.com/NixOS/rfcs/pull/137#discussion_r1050784664)